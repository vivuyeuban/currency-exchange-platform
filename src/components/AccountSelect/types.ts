import type { ChangeEvent } from 'react';
import type { TextFieldProps } from '@material-ui/core';
import type { TAccount } from '../../reducers/types';

export type TAccountSelectOnChangeProps = ChangeEvent<HTMLInputElement | HTMLTextAreaElement>;

export type TAccountSelectProps = TextFieldProps & {
    accounts: TAccount[];
    label: string;
};
