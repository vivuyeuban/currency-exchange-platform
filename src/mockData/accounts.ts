import type { TAccount } from '../reducers/types';

const accounts: TAccount[] = [
  {
    id: 'account-1',
    name: 'Dollars Account',
    currency: 'USD',
    balance: 200,
  },
  {
    id: 'account-2',
    name: 'Euros Account',
    currency: 'EUR',
    balance: 150,
  },
  {
    id: 'account-3',
    name: 'Pounds Account',
    currency: 'GBP',
    balance: 10,
  },
];

export default accounts;
