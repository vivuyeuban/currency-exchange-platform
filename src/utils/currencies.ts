export type TCurrencyStringConfig = {
  amount: string | number,
  precision?: number,
  currency?: string
};

export const getLocalString = ({
  amount,
  precision = 2,
  currency,
}: TCurrencyStringConfig): string => {
  const parsed = parseFloat(amount as string);

  if (Number.isNaN(parsed)) {
    return '';
  }

  return new Intl.NumberFormat('en-US', {
    style: currency ? 'currency' : 'decimal',
    currency,
    maximumFractionDigits: precision,
    minimumFractionDigits: precision,
  }).format(parsed);
};

export const stripNonDigits = (string: string): string => string.replace(/[^0-9.]/g, '');

export const getProperlyRounded = (value: string | number, precision = 2): string => {
  const newVal = typeof value === 'string' ? stripNonDigits(value) : value;
  if (!newVal) {
    return '';
  }
  const coef = 10 ** precision;
  return (Math.round(Number(newVal) * coef) / coef).toFixed(precision);
};

export const getCurrencySymbol = (currency: string): string => new Intl.NumberFormat('en-US', {
  style: 'currency',
  currency,
}).format(1)[0];

export const toCurrencyString = ({
  amount,
  precision = 2,
  currency,
}: TCurrencyStringConfig): string => {
  let reservedVal: string = stripNonDigits(String(amount));
  if (!reservedVal) {
    return '';
  }

  const hasDot = reservedVal.includes('.');
  const fractional = reservedVal.split('.')[1];
  let precisionLength = fractional?.length || 0;

  if (fractional && fractional.length > precision) {
    precisionLength = precision;
    reservedVal = getProperlyRounded(reservedVal, precision);
  }

  const converted = getLocalString({
    amount: reservedVal,
    precision: precisionLength,
    currency,
  });

  const postfix = (hasDot && !fractional) ? '.' : '';
  return `${converted}${postfix}`;
};
