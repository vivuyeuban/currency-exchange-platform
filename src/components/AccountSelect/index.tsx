import React from 'react';

import MenuItem from '@material-ui/core/MenuItem';
import TextField from '@material-ui/core/TextField';

import { toCurrencyString } from '../../utils/currencies';

import type { TAccountSelectProps } from './types';

const AccountSelect = React.memo(({
  accounts = [],
  label,
  ...props
}: TAccountSelectProps) => (
  <TextField
    {...props}
    label={label}
    variant="outlined"
    select
  >
    {
      accounts.map((account) => {
        const accountBalance = toCurrencyString({
          amount: account.balance.toString(),
          currency: account.currency,
        });
        const accountLabel = `${account.name} - ${accountBalance}`;
        return (
          <MenuItem key={account.id} value={account.id}>{accountLabel}</MenuItem>
        );
      })
    }
  </TextField>
));

export default AccountSelect;
