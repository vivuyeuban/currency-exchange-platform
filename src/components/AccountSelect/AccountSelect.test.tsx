import React from 'react';
import { fireEvent, render, within } from '@testing-library/react';
import AccountSelect from './index';
import mockData from '../../mockData/accounts';
import { toCurrencyString } from '../../utils/currencies';

test('renders and matches snapshot', () => {
  const { container } = render(<AccountSelect
    label="TestLabel"
    accounts={mockData}
    value={mockData[0].id}
  />);
  expect(container).toMatchSnapshot();
});

test('contains formatted value', () => {
  const { getByRole } = render(<AccountSelect
    label="TestLabel"
    accounts={mockData}
    value={mockData[0].id}
  />);
  const displayedValue = getByRole('button');
  const formattedBalance = toCurrencyString({
    amount: mockData[0].balance.toString(),
    precision: 2,
    currency: mockData[0].currency,
  });
  expect(displayedValue).toHaveTextContent(`${mockData[0].name} - ${formattedBalance}`);
});

test('fires onChange event with account id', async () => {
  const mockFn = jest.fn();
  const { getByRole } = render(<AccountSelect
    label="TestLabel"
    accounts={mockData}
    onChange={mockFn}
    value={mockData[0].id}
  />);
  fireEvent.mouseDown(getByRole('button'));

  const listbox = within(getByRole('listbox'));
  const accounts = listbox.getAllByRole('option');
  fireEvent.click(accounts[2]);

  expect(mockFn).toBeCalledTimes(1);
});
