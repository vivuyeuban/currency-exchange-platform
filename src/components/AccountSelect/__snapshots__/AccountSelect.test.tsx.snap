// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and matches snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
      data-shrink="true"
    >
      TestLabel
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
    >
      <div
        aria-haspopup="listbox"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
        role="button"
        tabindex="0"
      >
        Dollars Account - $200
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        tabindex="-1"
        value="account-1"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
        >
          <span>
            TestLabel
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
